include/unifex/unstoppable_baton.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+
+#include <atomic>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _ub {
+
+struct _op_base;
+
+template <typename Receiver>
+struct _operation {
+  struct type;
+};
+
+template <typename Receiver>
+using operation = typename _operation<Receiver>::type;
+
+struct unstoppable_baton;
+
+struct _sender {
+  template <template <class...> class Variant,
+            template <class...> class Tuple>
+  using value_types = Variant<Tuple<>>;
+
+  template <template <class...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+
+  static constexpr bool sends_done = false;
+
+  explicit _sender(unstoppable_baton& baton) noexcept
+    : baton_(&baton) {}
+
+  template (typename Receiver)
+    (requires receiver_of<Receiver>)
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) const noexcept(
+      std::is_nothrow_constructible_v<std::decay_t<Receiver>, Receiver&&>) {
+    return operation<std::decay_t<Receiver>>{*baton_, (Receiver&&)r};
+  }
+
+ private:
+  unstoppable_baton* baton_;
+};
+
+struct unstoppable_baton {
+  unstoppable_baton() noexcept
+    : unstoppable_baton(false) {}
+
+  explicit unstoppable_baton(bool startSignalled) noexcept
+    : state_(startSignalled ? this : nullptr) {}
+
+  void post() noexcept;
+
+  bool ready() const noexcept {
+    return state_.load(std::memory_order_acquire) ==
+        static_cast<const void*>(this);
+  }
+
+  void reset() noexcept {
+    // transition from signalled (i.e. state_ == this) to not-signalled
+    // (i.e. state_ == nullptr).
+    void* oldState = this;
+
+    // We can ignore the the result.  We're using _strong so it won't fail
+    // spuriously; if it fails, it means it wasn't previously in the signalled
+    // state so resetting is a no-op.
+    (void)state_.compare_exchange_strong(
+        oldState, nullptr, std::memory_order_acq_rel, std::memory_order_relaxed);
+  }
+
+  [[nodiscard]] _sender wait() noexcept {
```

Comment:
It would be nice to not use the word wait in non-blocking code. I would like 'wait' to become a code-smell. `sync_wait(b.wait())` is also awkward.

options:

```
sender on_ready()
sync_wait(b.on_ready())
```
[on_ is not my favorite prefix.. w.on_mouse_moves() or w.mouse_moves()?]

```
bool is_ready()
sender ready()
sync_wait(b.ready())
```

Reply:
I'm happy to change; your ideal of making `wait` a code smell sounds nice.  I'm uncertain about your suggested names, though.  Predicates in the STL seem not to start with `is_` (e.g. `vector::empty()`), so `bool is_ready()` and `sender ready()` feels suboptimal.

Reply:
I settled on `async_manual_reset_event`, `set`, `reset`, and `async_wait`.  I'm a little uncomfortable losing the "unstoppable" nature of the type since you _could_ write a stoppable version at some cost increment, and I'm a little concerned @kirkshoop won't be happy with `async_wait`, so feel free to suggest new names.